src/otaclient/_logging.py

```diff
@@ -19,32 +19,181 @@
+            )
+            _transmitter_grpc.check(timeout=3)
+            return _transmitter_grpc
+        except Exception:
+            return TransmitterHttp(
+                logging_upload_endpoint=logging_upload_endpoint, ecu_id=ecu_id
+            )
+
+
 class _LogTeeHandler(logging.Handler):
     """Implementation of teeing local logs to a remote otaclient-iot-logger server."""
 
     def __init__(self, max_backlog: int = 2048) -> None:
         super().__init__()
-        self._queue: Queue[str | None] = Queue(maxsize=max_backlog)
+        self._queue: Queue[QueueData | None] = Queue(maxsize=max_backlog)
 
     def emit(self, record: logging.LogRecord) -> None:
         with contextlib.suppress(Exception):
-            self._queue.put_nowait(self.format(record))
+            _log_type = getattr(record, "log_type", LogType.LOG)  # default to LOG
+            # if a message is log message, format the message with the formatter
+            # otherwise(metric message), use the raw message
+            _message = (
+                self.format(record) if _log_type == LogType.LOG else record.getMessage()
+            )
+            self._queue.put_nowait(QueueData(_log_type, _message))
+
+    def _wait_for_log_server_up(
+        self, logging_upload_endpoint: AnyHttpUrl, ecu_id: str
+    ) -> None:
+        """Wait for the logging server to be up.
+
+        This is a blocking call. It will keep trying to connect to the server until it succeeds.
+        """
+        while True:
+            try:
+                # send empty message via HTTP and check if logging server is up
+                # because HEAD is not supported by old iot-logger server
+                TransmitterHttp(logging_upload_endpoint, ecu_id).send(
+                    log_type=LogType.LOG, message="", timeout=1
+                )
+                break
+            except (
+                requests.HTTPError,
+                requests.ConnectionError,
+                ConnectionRefusedError,
+            ):
+                # ignore the such exceptions and continue
+                pass
+            time.sleep(1)
+
+    def start_upload_thread(
+        self,
+        logging_upload_endpoint: AnyHttpUrl,
+        logging_upload_grpc_endpoint: AnyHttpUrl | None,
+        ecu_id: str,
+    ) -> None:
+        LOG_TRANSMITTER_TIMEOUT_SEC = 3
 
-    def start_upload_thread(self, endpoint_url: str):
         log_queue = self._queue
         stop_logging_upload = Event()
 
         def _thread_main():
-            _session = requests.Session()
+            self._wait_for_log_server_up(
+                logging_upload_endpoint=logging_upload_endpoint, ecu_id=ecu_id
+            )
+            _transmitter = TransmitterFactory.create(
+                logging_upload_endpoint=logging_upload_endpoint,
+                logging_upload_grpc_endpoint=logging_upload_grpc_endpoint,
+                ecu_id=ecu_id,
+            )
```

Comment:
Before entering while loop, wait for log server wake up by HTTP **without** timeout.